Added V9240 energy metering chip driver

aquaforum opened the pull request:
Added V9240 energy metering chip driver

**Related issue (if applicable):** fixes # 22868

## Checklist:
  - [*] The pull request is done against the latest development branch
  - [*] Only relevant files were touched
  - [*] Only one feature/fix was added per PR and the code change compiles without warnings
  - [*] The code change is tested and works with Tasmota core ESP8266 V.2.7.8
  - [-] The code change is tested and works with Tasmota core ESP32 V.3.1.3.250302
  - [*] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).
  
[Datasheet](https://www.vangotech.com/uploadpic/163903940488.pdf)
There is a certain discrepancy in the calibration commands of this chip. I tried to do it according to the description of these commands, but then I could not calibrate accurately. Now the readings in the tasmota ui  correspond to the commercial metering device I have.

aquaforum committed:
Add driver code similar to the prototype

aquaforum committed:
they are talking to each other

aquaforum committed:
Added implementation of calibration commands

aquaforum committed:
continued work of calibration

aquaforum committed:
Maybe it works.

aquaforum committed:
Add V9240 driver

aquaforum committed:
Addendum to previous commit

aquaforum committed:
Post-merger control

arendst commented on tasmota/tasmota_xx2c_global/xnrg_interface.ino:
```diff
@@ -154,7 +154,11 @@ bool (* const xnrg_func_ptr[])(uint32_t) = {   // Energy driver Function Pointer
 
 #ifdef XNRG_33  // Reserved for use by xsns_02_analog.ino
-  &Xnrg33
+  &Xnrg33,
```
Do not expand!!

Use 25 instead by updating your `xnrg_34_v9240.ino`

arendst commented on tasmota/include/tasmota_template_legacy.h:
```diff
@@ -471,7 +471,10 @@ const uint16_t kGpioConvert[] PROGMEM = {
   AGPIO(GPIO_IEM3000_RX),     // IEM3000 Serial interface
   AGPIO(GPIO_ZIGBEE_RST),     // Zigbee reset
-  AGPIO(GPIO_DYP_RX)
+  AGPIO(GPIO_DYP_RX),
```
Read the header:

// ATTENTION: No additions are supported

So remove your changes to this file.

arendst commented on tasmota/berry/include/be_gpio_defines.h:
```diff
@@ -358,7 +358,10 @@ const be_const_member_t lv_gpio_constants[] = {
     { "ZIGBEE_RST", (int32_t) GPIO_ZIGBEE_RST },
     { "ZIGBEE_RX", (int32_t) GPIO_ZIGBEE_RX },
-    { "ZIGBEE_TX", (int32_t) GPIO_ZIGBEE_TX },
+    { "ZIGBEE_TX", (int32_t) GPIO_ZIGBEE_TX },    
```
Read the header of this file:

* The keys in the array belox must be in lexicographic order

So add your changes at the correct place.

aquaforum committed:
Change driver number 34 to 25

aquaforum committed:
Correction of other comments

arendst commented on tasmota/tasmota_xx2c_global/xnrg_interface.ino:
```diff
@@ -154,7 +154,11 @@ bool (* const xnrg_func_ptr[])(uint32_t) = {   // Energy driver Function Pointer
-  &Xnrg33
+  &Xnrg33,
+#endif
+
```
Remove below duplicate code

aquaforum committed:
Removed  duplicate code

arendst commented:
Pls provide your calibration values.

I don't want to use the filesystem to store calibration data as we already have provided these in the Settings Area. If they don't fit you'll need to add some calculation in your driver to make them fit.

Regarding reactive power calibration, how about skipping the driver's reactive power measurement and use the default Tasmota reactive power calculation based on U, I and P. Pls compare if the Tasmota calculation nears the measured one and if they do drop support for the chip's (calibrated) reactive power (which will be a mess anyway to find a calibrated reactive load).

aquaforum committed:
Adjusting the calibration procedure according to the behavior stated here. To the extent possible.
https://tasmota.github.io/docs/Power-Monitoring-Calibration/#calibration-procedure

Jason2866 commented on tasmota/berry/include/be_gpio_defines.h:
```diff
@@ -358,7 +360,7 @@ const be_const_member_t lv_gpio_constants[] = {
     { "ZEROCROSS", (int32_t) GPIO_ZEROCROSS },
     { "ZIGBEE_RST", (int32_t) GPIO_ZIGBEE_RST },
     { "ZIGBEE_RX", (int32_t) GPIO_ZIGBEE_RX },
-    { "ZIGBEE_TX", (int32_t) GPIO_ZIGBEE_TX },
```
remove added trailing whitespaces

aquaforum commented:
I have adjusted the calibration procedure in accordance with the [documentation](https://tasmota.github.io/docs/Power-Monitoring-Calibration/#calibration-procedure) as much as possible.
I removed the reactive power measurement altogether, I could not understand how it works in this chip. Now the reactive power is calculated by the tasmota.

My personal thoughts: yes, it works, but it does not allow to calibrate the indicators accurately, so accurately that it passes metrological verification. Sometimes you have to adjust quite a bit, in fact, you need to take and change the coefficient of the meter very slightly. And do this relative to the current value. I reset the energy meter value in the tasmota, connected a commercial meter in series with the chip, the readings of which I trust, so they worked for several days (a week) and then I compared their readings, then calculated how much I need to change the calibration coefficient to achieve the same readings. Now the error of my device is about 0.1% relative to the commercial meter.

aquaforum committed:
Removed added trailing whitespaces

arendst commented on tasmota/tasmota_xnrg_energy/xnrg_25_v9240.ino:
```diff
@@ -0,0 +1,720 @@
+    if(port != nullptr)
+    {
+        port->end();
+        Serial.end();
```
Do not close Serial.

arendst commented on tasmota/tasmota_xnrg_energy/xnrg_25_v9240.ino:
```diff
@@ -0,0 +1,720 @@
+
+static const int32_t header = 0x7d;
+static const int32_t ctrl_read = 0x1;
+static const int32_t ctrl_write = 0x2;
```
Use unique global names like v9240_header

aquaforum committed:
Fixing several small issues.
